File: mobile/kmp/ui/src/commonMain/kotlin/com/atls/hyperion/ui/theme/tokens/borders/Borders.kt
package com.atls.hyperion.ui.theme.tokens.borders

import com.atls.hyperion.ui.theme.tokens.colors.Colors

class Borders(colors: Colors) {
    val neutral: PressableBorderStates = Neutral(colors)
    val action: PressableBorderStates = Action(colors)
    val danger: PressableBorderStates = Danger(colors)
    val dropzone: BorderStates = Dropzone(colors)

    private class Neutral(colors: Colors) : PressableBorderStates {
        override val default = Border(BorderWidth.sm, BorderStyle.Solid, colors.surface.subtle)
        override val pressed = Border(BorderWidth.sm, BorderStyle.Solid, colors.surface.base)
        override val focused = Border(BorderWidth.sm, BorderStyle.Solid, colors.action.base)
        override val disabled = Border(BorderWidth.sm, BorderStyle.Solid, colors.surface.soft)
    }

    private class Action(colors: Colors) : PressableBorderStates {
        override val default = Border(BorderWidth.sm, BorderStyle.Solid, colors.action.base)
        override val pressed = Border(BorderWidth.sm, BorderStyle.Solid, colors.action.pressed)
        override val focused = Border(BorderWidth.md, BorderStyle.Solid, colors.action.base)
        override val disabled = Border(BorderWidth.sm, BorderStyle.Solid, colors.action.disabled)
    }

    private class Danger(colors: Colors) : PressableBorderStates {
        override val default = Border(BorderWidth.sm, BorderStyle.Solid, colors.status.error)
        override val pressed = Border(BorderWidth.sm, BorderStyle.Solid, colors.status.error)
        override val focused = Border(BorderWidth.md, BorderStyle.Solid, colors.status.error)
        override val disabled = Border(BorderWidth.sm, BorderStyle.Solid, colors.surface.soft)
    }

    private class Dropzone(colors: Colors) : BorderStates {
        override val default = Border(BorderWidth.sm, BorderStyle.Dashed, colors.surface.muted)
        override val focused = Border(BorderWidth.md, BorderStyle.Dashed, colors.action.base)
        override val disabled = Border(BorderWidth.sm, BorderStyle.Dashed, colors.surface.soft)
    }
}

File: ui/theme/src/tokens/borders/borders.ts
import type { Colors }        from '../colors/index.js'
import type { BordersTokens } from './interfaces.js'

import { borderStyles }       from './constants.js'
import { borderWidths }       from './constants.js'
import { createBorder }       from './utils/index.js'

export const borders = (colors: Colors): BordersTokens => ({
  neutral: {
    default: createBorder(borderWidths.sm, borderStyles.solid, colors.surface.subtle),
    hover: createBorder(borderWidths.sm, borderStyles.solid, colors.surface.muted),
    pressed: createBorder(borderWidths.sm, borderStyles.solid, colors.surface.base),
    focused: createBorder(borderWidths.sm, borderStyles.solid, colors.action.base),
    disabled: createBorder(borderWidths.sm, borderStyles.solid, colors.surface.soft),
  },
  action: {
    default: createBorder(borderWidths.sm, borderStyles.solid, colors.action.base),
    hover: createBorder(borderWidths.sm, borderStyles.solid, colors.action.hover),
    pressed: createBorder(borderWidths.sm, borderStyles.solid, colors.action.pressed),
    focused: createBorder(borderWidths.md, borderStyles.solid, colors.action.base),
    disabled: createBorder(borderWidths.sm, borderStyles.solid, colors.action.disabled),
  },
  danger: {
    default: createBorder(borderWidths.sm, borderStyles.solid, colors.status.error),
    hover: createBorder(borderWidths.sm, borderStyles.solid, colors.status.error),
    pressed: createBorder(borderWidths.sm, borderStyles.solid, colors.status.error),
    focused: createBorder(borderWidths.md, borderStyles.solid, colors.status.error),
    disabled: createBorder(borderWidths.sm, borderStyles.solid, colors.surface.soft),
  },
  dropzone: {
    default: createBorder(borderWidths.sm, borderStyles.dashed, colors.surface.muted),
    hover: createBorder(borderWidths.sm, borderStyles.dashed, colors.action.hover),
    focused: createBorder(borderWidths.md, borderStyles.dashed, colors.action.base),
    disabled: createBorder(borderWidths.sm, borderStyles.dashed, colors.surface.soft),
  },
})
